db/types.go

```diff
@@ -0,0 +1,183 @@
+// It is safe for concurrent writes, following an optimistic (OCC) strategy, detecting any write
+// conflicts and returning an error on commit, rather than locking the DB.
+// This can be used to wrap a write-optimized batch object if provided by the backend implementation.
+type DBWriter interface {
+	// Set sets the value for the given key, replacing it if it already exists.
+	// CONTRACT: key, value readonly []byte
+	Set([]byte, []byte) error
+
+	// Delete deletes the key, or does nothing if the key does not exist.
+	// CONTRACT: key readonly []byte
+	Delete([]byte) error
+
+	// Flushes pending writes and discards the transaction.
+	// TODO: maybe change to Flush() and follow WriteBatch semantics (ie. don't discard)
+	Commit() error
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// DBReadWriter is a transaction interface that allows both reading and writing.
+type DBReadWriter interface {
+	DBReader
+	DBWriter
+}
+
+// Iterator represents an iterator over a domain of keys. Callers must call Close when done.
+// No writes can happen to a domain while there exists an iterator over it, some backends may take
+// out database locks to ensure this will not happen.
+//
+// Callers must make sure the iterator is valid before calling any methods on it, otherwise
+// these methods will panic. This is in part caused by most backend databases using this convention.
+//
+// As with DBReader, keys and values should be considered read-only, and must be copied before they are
+// modified.
+//
+// Typical usage:
+//
+// var itr Iterator = ...
+// defer itr.Close()
+//
+// for ; itr.Valid(); itr.Next() {
+//   k, v := itr.Key(); itr.Value()
+//   ...
+// }
+// if err := itr.Error(); err != nil {
+//   ...
+// }
+type Iterator interface {
+	// Domain returns the start (inclusive) and end (exclusive) limits of the iterator.
+	// CONTRACT: start, end readonly []byte
+	Domain() (start []byte, end []byte)
+
+	// Valid returns whether the current iterator is valid. Once invalid, the Iterator remains
+	// invalid forever.
+	Valid() bool
+
+	// Next moves the iterator to the next key in the database, as defined by order of iteration.
+	// If Valid returns false, this method will panic.
+	Next()
+
+	// Key returns the key at the current position. Panics if the iterator is invalid.
+	// CONTRACT: key readonly []byte
+	Key() (key []byte)
+
+	// Value returns the value at the current position. Panics if the iterator is invalid.
+	// CONTRACT: value readonly []byte
+	Value() (value []byte)
+
+	// Error returns the last error encountered by the iterator, if any.
+	Error() error
+
+	// Close closes the iterator, relasing any allocated resources.
+	Close() error
+}
+
+// VersionSet specifies a set of existing versions
+type VersionSet interface {
+	// Last returns the most recent saved version, or 0 if none
+	Last() uint64
+	// Count returns the number of saved versions
+	Count() int
```

Comment:
This is true for all the interfaces, not just this one.

Reply:
I'm going to push back a bit on this. The `VersionSet` should ideally represent the version history of a DB at a specific point in time, and should always be in a valid state, so `Count` and `All` should not fail. The iterator to replace `All` will have a `Valid` method, but should also not encounter an error.

If a DB can't return a valid version history, it should return an error from the `Versions` method.

Also, TMU this interface won't be exposed to SDK users. If the behavior changes, adding error return values should be a non-breaking change.

Reply:
Are you cashing versions when calling `Versions()`? If not, then `Count` can error if a connection to the DB will get broken in the meantime.

Reply:
Yes, versions are cached so that a `VersionSet` can refer to a self-contained version history state. If the number of saved versions is expected to be very large, we could change that, but then there's a race hazard - so we must

* lock the version history until the `VersionSet` is closed, or 
* merge the version access methods into the `DBConnection` interface, so they can be individually synchronized, or
* do nothing and add a caveat that the data is not thread-safe, leaving sync to the caller

I don't see another option yet.